db/migrate/20170609214859_optional_required_notes_field.rb

```diff
@@ -0,0 +1,27 @@
+class OptionalRequiredNotesField < ActiveRecord::Migration
+
+  class Product < ActiveRecord::Base
+  end
+
+  def up
+    add_column :products, :user_notes_field_mode, :string, null: false, default: "hidden"
+
+    Product.find_each do |product|
+      mode = product.note_available_to_users? ? "optional" : "hidden"
+      product.update_column(:user_notes_field_mode, mode)
+    end
+
+    remove_column :products, :note_available_to_users
```

Comment:
In this case, I'm using `update_column`, so the only way it should fail would be something catastrophic (i.e. db connection being lost). And if it does fail, we're going to end up in a weird state.

I don't think you can put DDL statements into a transaction on mysql. One thing I miss about postgres is that it automatically wraps the migrations in a transaction, so if anything fails it can roll back back schema changes.